### DISCUSSION TITLE
Introduce a type of unit tests for "simulating" updating a real repo

### DISCUSSION THREAD

**Discussion by olegkrivtsov** [renovatebot/renovate#24799] [Suggest an Idea]
### What would you like Renovate to be able to do?

While implementing some new features for Renovate, I constantly experience the need of testing how my changes work against a real repo. For example: https://github.com/renovatebot/renovate/pull/11891#discussion_r721872058

I feel there should be some kind of unit tests which would allow to check how my code works against a "real" repo. This has the following advantages:
 - better reliability, since manual testing obviously can't be done all times
 - cover an additional functional level which is not covered by unit tests now

### If you have any ideas on how this should be implemented, please tell us here.

1. I think this test should be platform-independent. 
2. The test would run against a "fake" repo fixture which would be stored in a __fixtures__ directory as its subdirectory. This fake repo would be a minimum reporduction repo.
3. The test would try to "update" the repo by generating a JSON object describing what pull requests it would like to create. This JSON would mimic the log entries we have when we run `yarn start` on a real repo.
4. The JSON produced would be compared with a snapshot. 

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment]
I have the same goals, but a different idea.

I have been thinking about the idea of a dummy/test/synthetic platform we can add for similar testing. On any run you'd have:
- A directory of fixture files somewhere which represent the main branch of the repo (or maybe you use subrepos to represent branches if you want to test having branches existing)
- A JSON file which represents the issue/PR/etc state

To keep it simple you probably only support a single repo per run.

So you'd run something like `yarn start --platform=dummy <path to fixtures>`.

The init repo logic would be similar to our git test logic where it would take the files and then create a temporary git repo on the local file system using the fixtures files. After that Renovate would git clone from the temporary git repo (to its regular temporary clone location..) and all util/git operations would be reused.

Similarly, the repo "state" would be loaded from JSON file, which would allow you to have scenarios where branches did or didn't exist, etc. The dummy platform type would support findPr() ensureIssue() etc and keep its state updated.

Finally you'd want a way to verify, e.g. maybe the state is written to a JSON file every change so once you're done running you can treat that file as a snapshot and run expect statements on it.

@viceice I'm sure you've had thoughts on synthetic e2e testing too - wdyt?

**Comment by viceice** [comment]
I like the idea with the dummy platform. But we should do those as e2e tests. So we won't slow down normal unit tests.

The real git tests are a the slowest tests already.

**Comment by viceice** [comment]
Maybe we can inject the dummy platform on e2e tests dynamically, so we don't ship it on npm?

**Comment by rarkins** [comment]
I think we can start off experimenting with it manually and after that it should be possible to build e2e automated testing around it

**Comment by olegkrivtsov** [comment]
@rarkins I think the dummy platform is a perfect idea!

**Comment by olegkrivtsov** [comment]
I have another idea. Not sure if it is better or worse comparing to the dummy platform.

For example, if we want to test agains a "real" minimum reproduction repo: https://github.com/olegkrivtsov/renovate-test-repo/
we can do the following in our unit test:

1. Create a "fixture" by cloning the repo at certain state determined by the commit hash, e.g.: https://github.com/olegkrivtsov/renovate-test-repo/commit/e6422a851969a8bb1bf069d75f5a9de08d642765
2. Create a fake `config.json` and specify that repo to run against.
3. Run `yarn start` for that local cloned repository with some flag (e.g. `--test-real-repo`) indicating we want to get the snapshot in JSON format.
4. Capture the snapshot and execute jest expects against it.

Although this test is platform-specific, but it does what is required. The only change needed is creating the `--test-real-repo` flag and dumping the snapshot in JSON format. Not sure how easy to implement that snapshot though.

Another requirement for such tests to run is to keep test repos (fixtures) in some "safe" place, so nobody removes a test repo by mistake/intentionally.
